### ISSUE TITLE
Dispatcher.Start() should probably be blocking

### ISSUE THREAD

**Issue by bombela** [yarpc/yarpc-go#635]
Right now, dispatcher.Start() spawns the necessary background tasks and return. This means most application must have a `select {}` block right after. Instead, we should probably have a blocking Start() method (We could rename it of course). It is always possible to call Start() in a goroutine if one wishes.

We can even offer a pair of methods:
 - Start() never block, StartAndWait() always block.
 - Or: Start() always block, StartBackground() never block.

**Comment by blampe** [comment]
Start(true/false)?

**Comment by kriskowal** [comment]
I think we can solve graceful shutdown generally by making a cancellable root context and make `Lifecycle { Run(context.Context) }`, instead of having Stop at all.

**Comment by kriskowal** [comment]
Though this doesn’t solve the problem of coordinating the order of initialization, ensuring that some components are started before starting others.

**Comment by bombela** [comment]
Ok after some offline discussion with @kriskowal, there is an interesting relationship with #634.

Every objects, including the Dispatcher can implement the lifecycle interface. For convenience, we also add a method on the Dispatcher for automatic Stopping on context cancellation.

The dispatcher could be offering:
 - Start() -> start everything in the background, then return.
 - Stop() -> stop everything running in the background, then return.
 - Run(ctx Context) -> Start(), then on ctx cancellation it calls Stop(), then return.

**Comment by kriskowal** [comment]
Consider that the dispatcher manages initialization order. We ensure that we listen on all transport inbounds *last*. To do this, we currently synchronize Start() in batches. Hiding Start() behind Run() makes that difficult—we would need another method that would block until Started().

I do like backing WaitUntilStarted() and WaitUntilStopped() with channels that close, to capture better that state transitions only go one way, and go that way permanently. We could compose most things from Run(), WaitUntilStarted(), and WaitUntilStopped().

**Comment by abhinav** [comment]
Is this issue resolved by simply exposing a `Stopped() chan struct{}` on `Dispatcher` (or something similar)? 

```go
if err := dispatcher.Start(); err != nil {
    log.Fatal(err)
}
<-dispatcher.Stopped()
```

**Comment by abhinav** [comment]
Or even just a `WaitUntilStopped` but just on Dispatcher?

**Comment by breerly** [comment]
I think this item is worth a gander before 1.0 - I've assigned to @kriskowal to investigate his lifecycle idea over the next few days.

**Comment by kriskowal** [comment]
We have abandoned this idea. Start() and Stop() block, and a deferred Stop() can prevent the main thread from exiting prematurely, provided that we trap SIGINT and SIGTERM. We have other ways to ensure that graceful shutdown is in fact graceful. See #669 for an example.
